File: api-server/src/application/mappers/guide_mapper.py
"""Guide mapper for entity-DTO conversion."""

from src.application.dtos import GuideResponseDTO
from src.domain.entities import Guide


class GuideMapper:
    """Mapper for Guide entity to DTOs."""

    @staticmethod
    def to_response_dto(guide: Guide) -> GuideResponseDTO:
        """Convert Guide entity to GuideResponseDTO."""
        return GuideResponseDTO(
            id=guide.id.value,
            guide_type=guide.guide_type.value,
            title=guide.title.value,
            description=guide.description,
            metadata=guide.metadata,
            step_ids=[
                step_id.value
                for step_id in guide.step_ids
            ],
            created_by_user_id=(
                guide.created_by_user_id.value
                if guide.created_by_user_id
                else None
            ),
            is_public=guide.is_public,
            is_highlighted=guide.is_highlighted,
            language=guide.language.value,
            created_at=guide.created_at.isoformat(),
            updated_at=guide.updated_at.isoformat(),
        )

File: mobile/src/infrastructure/mappers/GuideMapper.ts
import { Guide } from '@domain/entities/Guide'
import type { GuideDto, GuideCreateRequest, GuideUpdateRequest } from '../api/dtos/GuideDto'

/**
 * Mapper for converting between Guide domain entity and API DTOs.
 *
 * NOTE: Domain entities set createdAt/updatedAt in constructor.
 * When mapping from API DTO to domain entity, timestamps will be reset to current time.
 * This is a known limitation - functional behavior is unaffected (all business data preserved).
 */
export class GuideMapper {
  /**
   * Convert API DTO to domain entity.
   * Reconstructs stepIds array from API response.
   * Note: timestamps will be reset to current time due to entity constructor behavior.
   */
  static toDomain(dto: GuideDto): Guide {
    const guide = new Guide(
      dto.id,
      dto.guideType,
      dto.title,
      dto.description ?? undefined,
      dto.createdByUserId,
      dto.isPublic,
      dto.isHighlighted,
      dto.metadata ?? undefined,
      dto.language ?? 'en',
      dto.totalDuration ?? undefined
    )

    // Reconstruct stepIds array
    dto.stepIds.forEach(stepId => {
      guide.addStep(stepId)
    })

    return guide
  }

  /**
   * Convert domain entity to API DTO format.
   * Used for caching and verification.
   */
  static toDto(guide: Guide): GuideDto {
    return {
      id: guide.id,
      guideType: guide.guideType,
      title: guide.title,
      description: guide.description ?? null,
      metadata: guide.metadata ?? null,
      stepIds: guide.stepIds,
      createdAt: guide.createdAt.toISOString(),
      updatedAt: guide.updatedAt.toISOString(),
      createdByUserId: guide.createdByUserId,
      isPublic: guide.isPublic,
      isHighlighted: guide.isHighlighted,
      language: guide.language,
      totalDuration: guide.totalDuration ?? null,
    }
  }

  /**
   * Convert to API create request format.
   */
  static toCreateRequest(
    guideType: string,
    title: string,
    description?: string,
    isPublic: boolean = false,
    metadata?: Record<string, unknown>,
    language: string = 'en'
  ): GuideCreateRequest {
    return {
      guideType,
      title,
      description: description ?? null,
      metadata: metadata ?? null,
      isPublic,
      language,
    }
  }

  /**
   * Convert guide to API update request format.
   * Only includes mutable fields.
   */
  static toUpdateRequest(guide: Guide): GuideUpdateRequest {
    return {
      title: guide.title,
      description: guide.description ?? null,
      metadata: guide.metadata ?? null,
      isPublic: guide.isPublic,
      isHighlighted: guide.isHighlighted,
      language: guide.language,
    }
  }
}
